File: src/2015/1/b/js/solution.js
const { readInputData } = require('../../../../util');

const input = readInputData(__dirname);

const increment = '(';
const decrement = ')';

const steps = input.split('');


const destination = steps
  .reduce((floor, instruction, pos) => {
    if (typeof floor === 'object') {
      return floor;
    }

    if (floor === -1) {
      return {
        basementAt: pos,
      };
    }

    if (instruction === increment) {
      return floor + 1;
    }

    if (instruction === decrement) {
      return floor - 1;
    }

    return floor;
  }, 0);

if (typeof destination === 'object') {
  console.log('Santa enters the basement at position:', destination.basementAt);
} else {
  console.log('Santa never enters the basement');
}

File: src/2015/1/b/c/solution.c
#include "../../../../util/c/io.h"

#define INCREMENT '('
#define DECREMENT ')'


int main() {
  char buffer = 0;
  int floor = 0;
  int count = 0;

  while (mread(&buffer, 1) > 0) {
    if (buffer == INCREMENT) {
      floor += 1;
    }

    if (buffer == DECREMENT) {
      floor -= 1;
    }

    count += 1;

    if (floor < 0) {
      break;
    }
  }

  mprintf("Santa enters the basement at position: %d\n", count);

  return 0;
}
